lib/png/png.go

```diff
@@ -0,0 +1,119 @@
+package png
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	_ "embed"
+
+	"github.com/playwright-community/playwright-go"
+	"oss.terrastruct.com/d2/lib/xmain"
+)
+
+type Playwright struct {
+	PW      *playwright.Playwright
+	Browser playwright.Browser
+	Page    playwright.Page
+}
+
+func (pw *Playwright) RestartBrowser() (newPW Playwright, err error) {
+	if err = pw.Browser.Close(); err != nil {
+		return Playwright{}, err
+	}
+	return startPlaywright(pw.PW)
+}
+
+func (pw *Playwright) Cleanup() error {
+	if err := pw.Browser.Close(); err != nil {
+		return err
+	}
+	if err := pw.PW.Stop(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func startPlaywright(pw *playwright.Playwright) (Playwright, error) {
+	browser, err := pw.Chromium.Launch()
+	if err != nil {
+		return Playwright{}, err
+	}
+	context, err := browser.NewContext()
+	if err != nil {
+		return Playwright{}, err
+	}
+	page, err := context.NewPage()
+	if err != nil {
+		return Playwright{}, err
+	}
+	return Playwright{
+		PW:      pw,
+		Browser: browser,
+		Page:    page,
+	}, nil
+}
+
+func InitPlaywright() (Playwright, error) {
+	// check if playwright driver/browsers are installed and up to date
+	// https://github.com/playwright-community/playwright-go/blob/8e8f670b5fa7ba5365ae4bfc123fea4aac359763/run.go#L64.
+	driver, err := playwright.NewDriver(&playwright.RunOptions{})
+	if err != nil {
+		return Playwright{}, err
+	}
+	_, err = os.Stat(driver.DriverBinaryLocation)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = playwright.Install()
+			if err != nil {
+				return Playwright{}, err
+			}
+		} else {
+			return Playwright{}, fmt.Errorf("could not access Playwright binary location: %v\nerror: %w\nplease report this issue here: https://github.com/terrastruct/d2/issues/new", driver.DriverBinaryLocation, err)
+		}
+	}
+
+	cmd := exec.Command(driver.DriverBinaryLocation, "--version")
+	output, err := cmd.Output()
+	if err != nil {
+		return Playwright{}, fmt.Errorf("error getting Playwright version: %w\nplease report this issue here: https://github.com/terrastruct/d2/issues/new", err)
+	}
+	if !bytes.Contains(output, []byte(driver.Version)) {
+		err = playwright.Install()
+		if err != nil {
+			return Playwright{}, err
+		}
+	}
+
+	pw, err := playwright.Run()
+	if err != nil {
+		return Playwright{}, err
+	}
+	return startPlaywright(pw)
+}
+
+//go:embed generate_png.js
+var genPNGScript string
+
+const pngPrefix = "data:image/png;base64,"
+
+func ConvertSVG(ms *xmain.State, page playwright.Page, svg []byte) (outputImage []byte, err error) {
+	encodedSVG := base64.StdEncoding.EncodeToString(svg)
+	pngInterface, err := page.Evaluate(genPNGScript, "data:image/svg+xml;charset=utf-8;base64,"+encodedSVG)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate png: %w\nplease report this issue here: https://github.com/terrastruct/d2/issues/new", err)
```

Comment:
We don't need `please report this issue here`. I've never understood the value of such notices and if we do want them, that's a separate PR and they should be centrally located in a constructor of some sort or when outputting errors.

Reply:
If someone can't find our issue tracker within 10 seconds, frankly I don't want them reporting an issue.

Reply:
it's not about discoverability. it's removing friction to behavior that's helpful to us. this is basic UX to not make users work. even if the work seems trivially easy, removing it is likely many useful reports that otherwise would not be made.

Reply:
agreed on the second point tho, let's just save these for separate PR that centralizes

Reply:
i disagree. endless bug reports are not behaviour that's helpful for us. i promise you, if people start using d2 and it gets even a thousand, they *will* start making issues left and right regardless of whether we show this message. we'll need to hire someone full time just to manage the repo, the questions and the discussions. it will snowball hard.

at which point this will just be noise. and without centralizing, it screws up go's `%w` fmt verb which has to be at the end of the error format string. we also have no indication that it's a mistake on our end, maybe it's something wrong with their setup. giving them the issue link for everything will make users open lazy issues and we don't have time for that.

Reply:
and linking directly to the new issue page means people won't search first, they'll just fire it off.

Reply:
how about we include it while it's small scale (prioritize visibility into issues), and then remove it when it's big?

Reply:
no i think it's useless but i'll defer as long as we remove it eventually.